pkg/accounting/accounting.go

```diff
@@ -923,10 +958,14 @@ func (a *Accounting) NotifyRefreshmentReceived(peer swarm.Address, amount *big.I
 }
 
 // PrepareDebit prepares a debit operation by increasing the shadowReservedBalance
-func (a *Accounting) PrepareDebit(peer swarm.Address, price uint64) (Action, error) {
+func (a *Accounting) PrepareDebit(ctx context.Context, peer swarm.Address, price uint64) (Action, error) {
 	accountingPeer := a.getAccountingPeer(peer)
 
-	accountingPeer.lock.Lock()
+	if err := accountingPeer.lock.TryLock(ctx); err != nil {
```

Comment:
I am bit confused, why cant we check if context is done directly inside PrepareDebit and Credit? Why the need for a custom mutex?

Reply:
We want to wait to acquire the lock but only for as long as the context permits us, this is why we have the select construct in the `TryLock` method.
Just checking for context expiration in the method would weed out some expired goroutines but we want to go further and reject those goroutines which were not able to get the lock within the deadline.

Reply:
I see, so we are essentially racing the context expiration with lock acquiral, which means we have seen cases where these functions were under lock contention, for longer periods than one second, judging from the code.